File: HW5/PolygonDyn.cpp
#include "PolygonDyn.h"
PolygonDyn::PolygonDyn()
{
	vecPoints = new point2d[1];
	setDynamicSize(1);
	vecPoints[0].x = 0;	/*There is a default constructor if it needed.*/
	vecPoints[0].y	= 0;	
}
PolygonDyn::PolygonDyn(vector<polygon::point2d> vec)
{	

	vecPoints = new point2d[2*(vec.size())+1];
	setDynamicSize(2*(vec.size())+1);	/*This gets a vector of point2d and keep in this points in the dynamic memory. */

	int i = 0;
	while(i<vec.size())
	{
		vecPoints[i].x = vec[i].x;
		vecPoints[i].y = vec[i].y;
		i++;
	}
	Area = 0;
}

PolygonDyn::PolygonDyn(point2d& point)
{
	vecPoints = new point2d[1];	/*Gets a point and keep them in the dynamic memory.*/
	setDynamicSize(1);
	vecPoints[0].x = point.x;
	vecPoints[0].y	= point.y;		
	Area = 0;
			
}

PolygonDyn::PolygonDyn(Rectangle& r)
{
	vecPoints = new point2d[4];
	//x1=x, y1=y, x2=x, y2=y+height, x3=x+width, y3=y+height, x4=x+width, y4=y
	setDynamicSize(4);	

	vecPoints[0].x = r.getX();
	vecPoints[0].y = r.getY();	/*Rectangle convertion constructor. Converts it a polygon.*/
	vecPoints[1].x = r.getX();
	vecPoints[1].y = r.getY()+r.getHeight();
	vecPoints[2].x = r.getX()+r.getWidth();
	vecPoints[2].y = r.getY()+r.getHeight();
	vecPoints[3].x = r.getX()+r.getWidth();
	vecPoints[3].y = r.getY();
	Area = (r.getWidth())*(r.getHeight());

}

PolygonDyn::PolygonDyn(Triangle& t)
{
	vecPoints = new point2d[3];
	setDynamicSize(3);

	vecPoints[0].x = t.getX1();/*Triangle convertion constructor.*/
	vecPoints[0].y = t.getY1();
	vecPoints[1].x = t.getX2();
	vecPoints[1].y = t.getY2();
	vecPoints[2].x = t.getX3();
	vecPoints[2].y = t.getY3();	
	Area = (t.getSide())*(t.getSide())*(pow(3,0.5))/4;
}

PolygonDyn::PolygonDyn(Circle& c)
{
	int i;
	int r = c.getRadius();/*Circle convertion constructor. Converts it a polygon.*/
	double a;


	vecPoints = new point2d[99];
	setDynamicSize(99);

	a=2*PI/99;

	for(i=0 ; i<99 ; i++)
	{	
		vecPoints[i].x = r*cos(a)+c.getX();
		vecPoints[i].y = r*sin(a)+c.getY();
		
		a+=(2*PI/99);
	}
	Area = PI*(c.getRadius())*(c.getRadius());
}

Shape& PolygonDyn::operator++(int dummy)
{
	PolygonDyn* temp;
	temp = new PolygonDyn;

	int i;
	for(i=0 ; i<dynamicArraySize ; i++)
	{
		temp->vecPoints[i].x = (vecPoints[i].x);
		temp->vecPoints[i].y = (vecPoints[i].y);

	}

	for(i=0 ; i<dynamicArraySize ; i++)
	{
		(vecPoints[i].x)++;
		(vecPoints[i].y)++;

	}
	return *temp;
}

Shape& PolygonDyn::operator--(int dummy)
{
	PolygonDyn* temp;
	temp = new PolygonDyn;

	int i;
	for(i=0 ; i<dynamicArraySize ; i++)
	{
		temp->vecPoints[i].x = (vecPoints[i].x);
		temp->vecPoints[i].y = (vecPoints[i].y);
		
	}

	for(i=0 ; i<dynamicArraySize ; i++)
	{
		(vecPoints[i].x)--;
		(vecPoints[i].y)--;
	
	}
	return *temp;
}

int PolygonDyn::getDynamicSize() const
{
	return dynamicArraySize;	/*A getter function to get size of dynamic array.*/
}
void PolygonDyn::setDynamicSize(int sizeD)
{
	dynamicArraySize = sizeD;		/*A setter function to give size of dynamic array.*/
}


PolygonDyn::point2d* PolygonDyn::getVecPoints() const
{
	return vecPoints;		/*A getter function to get pointer.*/
}

PolygonDyn::PolygonDyn(const PolygonDyn& plObj)
	 :dynamicArraySize(plObj.dynamicArraySize)
{
	int i;
	vecPoints = new point2d[dynamicArraySize];/*This is a copy constructor. It needed for big three.*/
	for (i=0 ; i<dynamicArraySize ; i++)
		vecPoints[i] = plObj.vecPoints[i];
}

PolygonDyn& PolygonDyn::operator =(const PolygonDyn& rightSide)
{
	int i;
	if (dynamicArraySize != rightSide.dynamicArraySize)	/*Assignment operator to create big three.*/
	{
		delete [] vecPoints;
		vecPoints = new point2d[rightSide.dynamicArraySize];
	}

	dynamicArraySize = rightSide.dynamicArraySize;
	for(i=0 ; i<dynamicArraySize ; i++)
		vecPoints[i] = rightSide.vecPoints[i];

	return * this;
}

PolygonDyn::~PolygonDyn() 
{ 
    delete []vecPoints ; 
} 

void PolygonDyn::draw(ostream& res)
{
	int i;

	res<<"\n<polygon points=\"";      /*This function necessary to draw the given polygon shape. */ 	

	for(i=0 ; i<dynamicArraySize ; i++)
			 res<<" "<<vecPoints[i].x<<" "<<vecPoints[i].y;
	
	res<<"\" stroke = \"black\" fill=\"red\"></polygon>";

	
}

ostream& PolygonDyn::operator<<(ostream& res)
{
	int i;

	res<<"\n<polygon points=\"";      /*This function necessary to draw the given polygon shape. */ 	

	for(i=0 ; i<dynamicArraySize ; i++)
			 res<<" "<<vecPoints[i].x<<" "<<vecPoints[i].y;
	
	res<<"\" stroke = \"black\" fill=\"red\"></polygon>";

	return res;
}


bool PolygonDyn::operator ==(const PolygonDyn& obj)
{	int i, count=0;
	if(getDynamicSize() == obj.getDynamicSize())
	{
		for(i=0 ; i<getDynamicSize() ; i++)
		{	
			if((vecPoints[i].x!=obj.vecPoints[i].x)||(vecPoints[i].y!=obj.vecPoints[i].y))
				count++;	
		}
	}
	if(count!=0)
		return false;
	else
		return true;
 /*Boolean operators to control equality.*/
} 

bool PolygonDyn::operator !=(const PolygonDyn& obj)
{
	int i, count=0;
	if(getDynamicSize() != obj.getDynamicSize())
			return true;
	if(getDynamicSize() == obj.getDynamicSize())
	{
		for(i=0 ; i<getDynamicSize() ; i++)
		{	
			if((vecPoints[i].x!=obj.vecPoints[i].x)||(vecPoints[i].y!=obj.vecPoints[i].y))
				return true;	
		}
	}

	return false;
 /*Boolean operators to control equality.*/
} 

polygon::point2d PolygonDyn::operator [](int index)
{
    if (index>dynamicArraySize || index<0)
        exit(0);
	return vecPoints[index];
}
/*Index operators gives the value of given index.*/

const polygon::point2d PolygonDyn::operator [](int index) const
{
    if (index>dynamicArraySize || index<0)
        cout<<"This index is invalid";
	return vecPoints[index];
}

File: HW7 (JAVA)/PolygonDyn.java
import java.awt.Graphics;


 public class PolygonDyn extends Polygon{

	private	point2d[] vecPoints;

	public PolygonDyn()
	{ }


	PolygonDyn(double a, double b)
	{
		point2d p = new point2d();

		vecPoints = new point2d[1];	
		p.x = a;
		p.y = b;
		vecPoints[0] = p;
		
		Area = 0;	
		Perimeter = 0;	
	}
	/**
	 * Takes a rectangle and convert to a polygon.
	 * @param r Rectangle object.
	 */
	PolygonDyn(Rectangle r)
	{
		vecPoints = new point2d[4];

		point2d p = new point2d();
		p.x = (r.getX());
		p.y = (r.getY());
		vecPoints[0] = p;

		p = new point2d();
		p.x = (r.getX());
		p.y = (r.getY()+r.getHeight());
		vecPoints[1] = p;
	
		p = new point2d();
		p.x = (r.getX()+r.getWidth());
		p.y = (r.getY()+r.getHeight());
		vecPoints[2] = p;

		p = new point2d();
		p.x = (r.getX()+r.getWidth());
		p.y = (r.getY());
		
		vecPoints[3] = p;
		
		Area = (r.getWidth())*(r.getHeight());
		Perimeter = 2*(r.getWidth()+r.getHeight());
	}
	/**
	 * Takes a triangle and convert to a polygon.
	 * @param t triangle object.
	 */
	PolygonDyn(Triangle t)
	{
		vecPoints = new point2d[3];
		point2d p = new point2d();
		
		p.x = (t.getX1());
		p.y = (t.getY1());
		
		vecPoints[0] = p;

		p = new point2d();
		p.x = (t.getX2());
		p.y = (t.getY2());
		
		vecPoints[1] = p;
	
		p = new point2d();

		p.x = (t.getX3());
		p.y = (t.getY3());
		
		vecPoints[2] = p;
		
		Area = (t.getSide())*(t.getSide())*(Math.pow(3,0.5))/4;
		Perimeter = 3*(t.getSide());		
	}
	/**
	 * Takes a circle and convert to a polygon.
	 * @param c circle object.
	 */
	PolygonDyn(Circle c)
	{
		int i;
		int r = c.getRadius();
		double a;

		vecPoints = new point2d[99];
		point2d p = new point2d();

		a=2*Math.PI/99;

		for(i=0 ; i<99 ; i++)
		{	
			p = new point2d();
			p.x = (r*Math.cos(a)+c.getX());
			p.y = (r*Math.sin(a)+c.getY());
			
			vecPoints[i] = p;
			
			a+=(2*Math.PI/99);
		}
		Area = Math.PI*(c.getRadius())*(c.getRadius());	
		Perimeter = 2*Math.PI*(c.getRadius());	
	}	
	/**
	 * 
	 * @return area of the shape
	 */
	public double area()
	{ return Area;}
	/**
	 * 
	 * @return perimeter of the shape
	 */
	public double perimeter()
	{ return Perimeter;}
	/**
	 * incremets the coordinates
	 */
	public void increment()
	{
		int i;
		for(i=0 ; i<vecPoints.length ; i++)
		{
			vecPoints[i].x = ((vecPoints[i].getX())+1);
			vecPoints[i].y = ((vecPoints[i].getY())+1); 
		}
	}
	/**
	 * decrements the coordinates
	 * 
	 */
	public void decrement()
	{
		int i;
		for(i=0 ; i<vecPoints.length ; i++)
		{
			vecPoints[i].x = ((vecPoints[i].getX())-1);
			vecPoints[i].y = ((vecPoints[i].getY())-1);
		}
	}
	
	/**
	 * This method draws the shapes.
	 * @param g Graphics object for drawing
	 */
	public void draw(Graphics g)
	{
		
		int[] xPoints = new int[vecPoints.length];
		int[] yPoints = new int[vecPoints.length];

		int i;
		for(i=0 ; i<vecPoints.length ; i++)
		{
			xPoints[i] = (int)(vecPoints[i].x);
			yPoints[i] = (int)(vecPoints[i].y);
			
		}
		
		g.fillPolygon(xPoints, yPoints, vecPoints.length); 
		
	}
	/**
	 * compares two shapes with recpect to their areas
	 * @param obj
	 */
	public int compareTo(Shape obj)
	{ 
		if(obj.area()>area())
			return 1;
		else if(obj.area()<area())
			return -1;
		else
			return 0;
	}

};


